File: domo/domo.php
<?php
include('/opt/PvMonit/function.php');

// Chargement de la config
$config = getConfigYaml('/opt/PvMonit');

function xml_data_get($DATA_FILE)  {
    global $config;
    $xmlData = null;
    try {
       
        $devices = new SimpleXMLElement(file_get_contents($DATA_FILE));
        foreach ($devices as $device) {
            //~ // On vérifie que la donnée ne soit pas périmé
            if ($device->timerefresh+$config['domo']['xmlDataExpir'] > time()) {
                foreach ($device->datas->data as $data) {
                    foreach ($config['domo']['valueUse'] as $id => $regexCheck) {
                        if ($data['id'] == $id) {
                            $xmlDataTemp = json_decode(json_encode($data->value), true);
                            trucAdir(5, 'XML parse : la valeur pour '.$data['id'].' à été trouvé à '.$xmlDataTemp[0]);
                            if (preg_match_all('/'.$regexCheck.'/', $xmlDataTemp[0])) {
                                $xmlData[$id] = $xmlDataTemp[0];
                            } else {
                                trucAdir(5, 'XML parse ERROR : La vérification regex pour '.$data['id'].' n\'est pas correct pour la valeur '.$xmlDataTemp[0].' ('.$regexCheck.')');
                                $xmlData[$id] = false;
                            }
                        }
                    }
                }
            } else {
                $xmlData = false;
                trucAdir(2, 'Les données sont périmées');
            }
        }
        // On vérifie si toutes les données sont là
        if (count($xmlData) < count($config['domo']['valueUse'])) {
            trucAdir(5, count($xmlData).' éléments sont trouvés alors que '.count($config['domo']['valueUse']).' éléments sont attendu dans la configuration');
            trucAdir(2, 'Toutes les données requisent ne sont pas présentes dans le XML donc on passe notre chemin (vérifier domo/valueUse dans le fichier config.yaml)');
            $xmlData = false;
        }
    } catch (Exception $e ) {
        $xmlData = false;
        trucAdir(2, 'Impossible de lire l\'XML : '.$e);
    }
    
    return $xmlData;
}

function MpptAbsOrFlo($cs, $timeUpNoBago = 10) {
    global $config;
    if ($timeUpNoBago == 0) {
        if (preg_match_all('/^Absorption|^Float/', $cs)) {
            return true;
        } else {
            return false;
        }
    } else {
        $fileCS=$config['domo']['prefixFileData'].'MpptFlo_NoBago'.'_CS_AoF';
        $fileTimerNoBago=$config['domo']['prefixFileData'].'MpptFlo_NoBago'.'_timer';
        if (preg_match_all('/^Absorption|^Float/', $cs)) {
            touch($fileCS);
            //~ echo "abs ou float";
            return true;
        } else {
            // Si j'étais en abs ou float j'allume le timer
            if (is_file($fileCS)) {
                $timer=time();
                file_put_contents($fileTimerNoBago, $timer);
                unlink($fileCS);
                //~ echo "création du timer/suppression du float abs";
                return true;
            // Si le timer est déjà en route on récupère l'info
            } else if (is_file($fileTimerNoBago)) {
                $timer=file_get_contents($fileTimerNoBago);
                // Le timer à été dépassé
                if (time() > $timer+$timeUpNoBago) {
                    //~ echo "timer dépassé";
                    unlink($fileTimerNoBago);
                    return false;
                } else {
                    //~ echo "timer NON dépassé";
                    return true;
                }
            // Sinon c'est que le régulateur n'est pas encore passé en abs/float
            } else {
                //~ echo "régulateur pas encore été en float abs";
                return false;
            }   
        }
    }
}

function MpptFlo($cs, $timeUpNoBago = 10) {
    global $config;
    if ($timeUpNoBago == 0) {
        if (preg_match_all('/^Float/', $cs)) {
            return true;
        } else {
            return false;
        }
    } else {
        $fileCS=$config['domo']['prefixFileData'].'MpptFlo_NoBago'.'_CS_AoF';
        $fileTimerNoBago=$config['domo']['prefixFileData'].'MpptFlo_NoBago'.'_timer';
        if (preg_match_all('/^Float/', $cs)) {
            touch($fileCS);
            //~ echo "abs ou float";
            return true;
        } else {
            // Si j'étais en abs ou float j'allume le timer
            if (is_file($fileCS)) {
                $timer=time();
                file_put_contents($fileTimerNoBago, $timer);
                unlink($fileCS);
                //~ echo "création du timer/suppression du float abs";
                return true;
            // Si le timer est déjà en route on récupère l'info
            } else if (is_file($fileTimerNoBago)) {
                $timer=file_get_contents($fileTimerNoBago);
                // Le timer à été dépassé
                if (time() > $timer+$timeUpNoBago) {
                    //~ echo "timer dépassé";
                    unlink($fileTimerNoBago);
                    return false;
                } else {
                    //~ echo "timer NON dépassé";
                    return true;
                }
            // Sinon c'est que le régulateur n'est pas encore passé en abs/float
            } else {
                //~ echo "régulateur pas encore été en float abs";
                return false;
            }   
        }
    }
}

trucAdir(4, 'Lancement du script');

if (!is_file($config['domo']['jsonFile']['etatPath'])) {
	genDefaultJsonFile('etat');
}
if (!is_file($config['domo']['jsonFile']['modPath'])) {
	genDefaultJsonFile('mod');
}


$initDb=false;
if (!is_file($config['domo']['dbFile'])) {
    $initDb=true;
}
// Connect DB
try {
    $dbco = new PDO('sqlite:'.$config['domo']['dbFile']);
	$dbco->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);
} catch ( PDOException $e ) {
	die('Connexion à la base : '.$e->getMessage());
}
// Create DB if not exists
if ($initDb==true) {
    try {
        $create = $dbco->query("CREATE TABLE relay (id INTEGER PRIMARY KEY, 
                                                      relay_number INTEGER, 
                                                      info VARCHAR(2), 
                                                      valeur INTEFER, 
                                                      date INTEGER, 
                                                      event TEXT)");
    } catch ( PDOException $e ) {
        echo 'Error initializing tables. Please contact the admin';
        die();
    }
}

function logInDb($relay, $info, $valeur, $event) {
    global $dbco;
    try {
		$insertcmd = $dbco->prepare("INSERT INTO relay (relay_number, info, valeur, date, event) 
										VALUES (:relay, :info, :valeur, ".time().", :event)");
		$insertcmd->bindParam('relay', $relay, PDO::PARAM_INT);
        $insertcmd->bindParam('info', $info, PDO::PARAM_STR);
		$insertcmd->bindParam('valeur', $valeur, PDO::PARAM_INT);
		$insertcmd->bindParam('event', $event, PDO::PARAM_STR);
		$insertcmd->execute();
	} catch ( PDOException $e ) {
		echo "DB error :  ", $e->getMessage();
		die();
	}
}


function relayLastUp($relay) {
    global $dbco;
    return  $dbco->query("SELECT date FROM relay WHERE relay_number = ".$relay." AND info = 'E' AND valeur = 1 ORDER BY date DESC LIMIT 1")->fetchColumn();        
}

function relayLastUpAuto($relay) {
    global $dbco;
    return  $dbco->query("SELECT date FROM relay WHERE relay_number = ".$relay."  AND info = 'M' AND valeur = 2 ORDER BY date DESC LIMIT 1")->fetchColumn();  
}

function relayLastDown($relay) {
    global $dbco;
    return  $dbco->query("SELECT date FROM relay WHERE relay_number = ".$relay." AND info = 'E' AND valeur = 0 ORDER BY date DESC LIMIT 1")->fetchColumn();  
}

# Est-ce que le relay c'est allumé puis est maintenant éteind aujourd'hui ? (dans les 12 heures)
function relayUpDownToday($relay) {
    global $dbco;
    $result = $dbco->query("SELECT count(date) FROM relay WHERE relay_number = ".$relay." AND info = 'M' AND (valeur = 1 OR valeur = 2) AND date > ".(time()-43200)." ORDER BY date DESC LIMIT 2")->fetchColumn();  
    if ($result == 2) {
        return true;
    } else {
        return false;
    }
}
# Est-ce que le relay c'est allumé aujourd'hui ? (dans les 12 heures)
function relayUpToday($relay) {
    global $dbco;
    $result = $dbco->query("SELECT count(date) FROM relay WHERE relay_number = ".$relay." AND info = 'E' AND valeur = 1 AND date > ".(time()-43200)." ORDER BY date DESC LIMIT 1")->fetchColumn();  
    if ($result >= 1) {
        return true;
    } else {
        return false;
    }
}
function timeUpMax($relay, $timeUp) {
    if (relayLastUpAuto($relay)+$timeUp < time()) {
        return true;
    } else {
        return false;
    }
}
function timeUpMin($relay, $timeUp) {
    if (relayLastUpAuto($relay)+$timeUp > time()) {
        return true;
    } else {
        return false;
    }
}


# Fonctions Timer
# Contrib @akoirium 
function timerStart($name, $time) {
	global $config;
    trucAdir(3, 'Début du timer '.$name.' pour '.$time.'s');
	file_put_contents($config['domo']['prefixFileData'].'timer_'.$name, time()+$time);
}
function timerOn($name) {
	global $config;
    if (!is_file($config['domo']['prefixFileData'].'timer_'.$name)) {
        // Le timer n'a pas été lancé, donc est pas terminé....
        trucAdir(2, 'Timer '.$name.' inconnu, on envoi TRUE pour dire que c\'est terminé...');
		return true;
	} else {
        $endTime=file_get_contents($config['domo']['prefixFileData'].'timer_'.$name);
        if ($endTime < time()) {
            // Timer On
            trucAdir(3, 'Timer '.$name.' pas terminé, on envoi FALSE');
            return false;
        } else {
            // Timer Off 
            trucAdir(3, 'Timer '.$name.' terminé, on envoi TRUE et on supprime le fichier');
            @unlink($config['domo']['prefixFileData'].'timer_'.$name);
            return true;
        }
    }
}


$dataCheckTime=0;
$xml_check_error=0;
$downloadNow=false;
$dataRefresh=false;
$relay_script_last_exec = 0;
$lastRefreshMod = 0;
$lastRefreshEtat = 0;
$dernierScriptJoue = 0;
while(true) {
    // Gestion du XML
    if (!is_file($config['tmpFileDataXml'])) {
        trucAdir(2, "Fichier inexistant");
        $downloadNow=true;    
    } else if (filemtime($config['tmpFileDataXml'])+$config['domo']['fileExpir'] < time()) {
        trucAdir(2, "Le fichier data est périmé !");
        $downloadNow=true;
    } elseif ($dataCheckTime+$config['domo']['dataCheckTime'] < time()) {
        trucAdir(4, "Préparation du rafraichissement des données");
        if (filemtime($config['tmpFileDataXml'])+$config['domo']['dataCheckTime'] > time()) {
            trucAdir(5, "Pas de téléchargement, le fichier semble très résent (peut être téléchargé par un autre script... lcd.py ?)"); 
            $dataRefresh=true;
        } else {
            $downloadNow=true;
        }
    } else {
        if ($xml_check_error == 0) {
            $downloadNow=false;
        }
    }

    // Téléchargement
    if ($downloadNow == true) {
        trucAdir(2, "Téléchargement du fichier xml");
        if ($xml_check_error != 0) {
            trucAdir(2, "C'est suite à une erreur, on patiente un peu...");
            sleep(10);
        }
        $opts = array('http' =>
            array(
                'method'  => 'GET',
                'timeout' => 60
            )
        );
        $context  = stream_context_create($opts);
        $result = file_get_contents($config['urlDataXml'], false, $context);
        file_put_contents($config['tmpFileDataXml'], $result);
        $dataRefresh=true;
        $downloadNow=false;
    }
    
    if ($dataRefresh==true) {
        trucAdir(4, "Rafraichissement des données");

        $xml_data_get=xml_data_get($config['tmpFileDataXml']);
        if ($xml_data_get == null || $xml_data_get == false) {
            trucAdir(2, 'Données XML invalide');
            $xml_check_error++;
            $downloadNow=true;
        } else {
            $xml_data_error=0;
            foreach ($xml_data_get as $id => $xml_data) {
                if ($xml_data === false) {
                    trucAdir(5, 'ERROR donnée '.$id.' est à false');
                    $xml_data_error++;
                }
            }
            if ($xml_data_error != 0) {
                trucAdir(4, 'ERROR Certaine données (dans config.yaml : valueUse) ne sont pas conforme ou pas présente.');
                $xml_check_error++;
                $downloadNow=true;
            } else {
                trucAdir(4, 'Les données sont bonnes !');
                $xml_check_error=0;
                $dataCheckTime=time();
            }
        }
        $dataRefresh=false;
    }
    
    // Sécurité, si pas de data pendant trop de temps, on éteind les relay en mod haut
    if ($dataCheckTime+$config['domo']['relay']['secuDownNoData'] < time()) {
        trucAdir(5, "SECURITE, trop de temps c'est écoulé sans donnée, on passe les relays en mode 2 (auto on) à 1 (auto off)");
        $relayModSecu = json_decode(file_get_contents($config['domo']['jsonFile']['modPath']), true);
        foreach ($relayModSecu as $relay => $Mod) {
            if ($Mod == 2) {
                $relayModSecu[$relay] = 1;
            }
        }
        file_put_contents($config['domo']['jsonFile']['modPath'], json_encode($relayModSecu));
    }
    
    // Récupération et enregistrement des données de mod
    if (filemtime($config['domo']['jsonFile']['modPath']) > $lastRefreshMod) {
        trucAdir(4, "Récupération des mods des relay");
        $relayMod_New = json_decode(file_get_contents($config['domo']['jsonFile']['modPath']), true);
        if (isset($relayMod)) {
            foreach (array_diff_assoc($relayMod_New,$relayMod) as $relayIdDiff => $relayModDiff) {
                trucAdir(4, "Enregistrement du changement sur le relay ".$relayIdDiff." vers le mod ".$relayModDiff);
                if (isset($saveInDb[$relayIdDiff]) && !is_null($saveInDb[$relayIdDiff])) {
                    trucAdir(4, "Avec le log ".$saveInDb[$relayIdDiff]['log']);
                    logInDb($relayIdDiff, 'M', $relayModDiff, $saveInDb[$relayIdDiff]['log']);
                    $saveInDb[$relayIdDiff]=null;
                } else {
                    logInDb($relayIdDiff, 'M', $relayModDiff, '');
                }
            }
        }
        $relayMod = $relayMod_New;
        $lastRefreshMod=time();
    }
    // Récupération et enregistrement des données d'état
    if (filemtime($config['domo']['jsonFile']['etatPath']) > $lastRefreshEtat) {
        trucAdir(4, "Récupération de l'état des relay");
        $relayEtat_New = json_decode(file_get_contents($config['domo']['jsonFile']['etatPath']), true);
        if (isset($relayEtat)) {
            foreach (array_diff_assoc($relayEtat_New,$relayEtat) as $relayIdDiff => $relayEtatDiff) {
                trucAdir(4, "Enregistrement du changement sur le relay ".$relayIdDiff." vers l'état ".$relayEtatDiff);
                logInDb($relayIdDiff, 'E', $relayEtatDiff, '');
            }
        }
        $relayEtat = $relayEtat_New;
        $lastRefreshEtat=time();
    }
    
    //
    // Traitement, ordre...
    //
    if ($xml_check_error == 0) {
        if ($relay_script_last_exec+$config['domo']['relay']['scriptExecInterval'] < time()) {
            
            
            if ($config['md5sum'] != md5_file('/opt/PvMonit/config.yaml')) {
                trucAdir(1, "Changement dans la configuration, relecture !");
                $config = getConfigYaml('/opt/PvMonit');
            }
            
            trucAdir(3, "Traitement des ordres");
            if (is_array($relayMod) && is_array($relayEtat)) {
                foreach ($relayMod as $relay => $Mod) {
                    //~ if ($relay > $dernierScriptJoue) {
                        // On s'occupe uniquement de ceux qui sont en mode auto
                        if ($Mod == 1 || $Mod == 2) {
                            if (is_file($config['domo']['relay']['scriptDir'].'/'.$relay.'.php')) {
                                $thisId=$relay;
                                $thisEtat=$relayEtat[$relay];
                                $thisMod=$relayMod[$relay];
                                $data=$xml_data_get;
                                $script_return = (include $config['domo']['relay']['scriptDir'].'/'.$relay.'.php');
                                if ($relay == 7) {
                                    //~ var_dump($relayMod[$relay]);
                                    //~ var_dump($script_return);
                                    trucAdir(2, "7 debug : ".$relayMod[$relay]."!=".$script_return['mod']);
                                }
                                if ($relayMod[$relay] != $script_return['mod'] && $script_return != null) {
                                    trucAdir(1, "Changement de mod pour le relay ".$relay." (".$thisMod." vers ".$script_return['mod'].")");
                                    trucAdir(2, "Pourquoi ? : ".$script_return['log']);
                                    $relayModPut=$relayMod;
                                    $relayModPut[$relay] = $script_return['mod'];
                                    file_put_contents($config['domo']['jsonFile']['modPath'], json_encode($relayModPut));
                                    $saveInDb[$relay]['mod'] = $script_return['mod'];
                                    $saveInDb[$relay]['log'] = $script_return['log'];
                                    $dernierScriptJoue = $relay;
                                    break;
                                } else {
                                    trucAdir(3, "Aucun changement d'état pour le relay ".$relay);
                                    if ($script_return['log'] != null) {
                                        trucAdir(3, '['.$thisId.']'.$script_return['log']);
                                    }
                                }
                            } else {
                                trucAdir(5, "Pas de script pour le relay ".$relay);
                            }
                        } else {
                            trucAdir(5, "Le relay ".$relay." n'est pas en mod automatique");
                        }
                        
                    //~ }
                    //~ $dernierScriptJoue = $relay;
                }
                //~ if ($dernierScriptJoue >= $config['domo']['relayNb']) {
                    //~ $dernierScriptJoue = 0;
                //~ }
            } else {
                trucAdir(3, "Aucune donnée sur les relay exploitable pour pouvoir lancer des actions");
            }
            $relay_script_last_exec=time();
        }
    }
    
    
    
    sleep(1);

}

?>

File: domo/domo.py
import yaml
import time
from smbus2 import SMBus
import os
from lxml import etree
from urllib.request import urlopen
import wget
from past.builtins import execfile
import re
import time
import sqlite3
import sys
import json


## for debug :
import pprint

with open('../config-default.yaml') as f1:
    config = yaml.load(f1, Loader=yaml.FullLoader)
with open('../config.yaml') as f2:
    config_perso = yaml.load(f2, Loader=yaml.FullLoader)

def configGet(key1, key2=None, key3=None, key4=None):
    if key4 != None:
        try:
            return config_perso[key1][key2][key3][key4]
        except:
            return config[key1][key2][key3][key4]
    elif key3 != None:
        try:
            return config_perso[key1][key2][key3]
        except:
            return config[key1][key2][key3]
    elif key2 != None:
        try:
            return config_perso[key1][key2]
        except:
            return config[key1][key2]
    else:
        try:
            return config_perso[key1]
        except:
            return config[key1]

# Cherche a savoir si le MPPT est en Absorption ou en Float (en fin de charge)
def MpptAbsOrFlo(cs):
    patternAbsFlo = re.compile(r"Absorption|Float")
    if patternAbsFlo.match(cs):
        return True;
    else : 
        return False;
def MpptFlo(cs):
    patternAbsFlo = re.compile(r"Float")
    if patternAbsFlo.match(cs):
        return True;
    else : 
        return False;

# Function for log
def logMsg(level, msg):
    if level <= configGet('printMessage') :
        print(time.strftime ('%m/%d/%Y %H:%M') ," - ",msg)
    return -1

# i2c write
def writeNumber(value):
    bus.write_byte(configGet('domo', 'i2c', 'adress'), value)
    return -1

def download_data():
    # téléchargement des données
    logMsg(3, 'Download data')
    with open(configGet('tmpFileDataXml'), 'wb') as tmpxml:
        tmpxml.write(urlopen(configGet('urlDataXml')).read())
    return time.time()

logMsg(1, 'Lancement du script domo.py')

heartGo=False
heartLastCheck=0
relayDataLastCheck=0
scriptExecLast=0
xmlLastCheck=0
xmlfileCheckError=0
xmlData = {}
xmlDataLastControleOk=0
xmlDataControle=False
dernierScriptJoue=-1
sortie=False
firstDataRece=True
bus=SMBus(configGet('domo', 'i2c', 'device'));

# BD
if not os.path.isfile(configGet('domo', 'dbFile')): 
    con = sqlite3.connect(configGet('domo', 'dbFile'))
    db = con.cursor()
    # Création de la base si elle n'existe pas 
    db.execute('''CREATE TABLE relay (id INTEGER PRIMARY KEY, relay_number INTEGER, info TEXT, valeur INTEFER, date INTEGER, event TEXT)''')
    con.commit()
else:
    con = sqlite3.connect(configGet('domo', 'dbFile'))
    db = con.cursor()

def logInDb(relay, info, valeur, event):
    db.execute("INSERT INTO relay VALUES (null,"+str(relay)+",'"+str(info)+"',"+str(valeur)+"," + str(int(time.time())) + ",'"+str(event)+"')")
    con.commit()

def relayLastUp(relay):
    req=db.execute('SELECT date FROM relay WHERE relay_number = '+str(relay)+' AND info = "E" AND (valeur = 2 OR valeur = 3) ORDER BY date DESC LIMIT 1')
    try:
        return req.fetchone()[0]
    except:
        return 0

def relayLastUpAuto(relay):
    req=db.execute('SELECT date FROM relay WHERE relay_number = '+str(relay)+' AND info = "E" AND valeur = 2 ORDER BY date DESC LIMIT 1')
    try:
        return req.fetchone()[0]
    except:
        return 0

def relayLastDown(relay):
    req=db.execute('SELECT date FROM relay WHERE relay_number = '+str(relay)+' AND info = "E" AND (valeur = 0 OR valeur = 1) ORDER BY date DESC LIMIT 1')
    try:
        return req.fetchone()[0]
    except:
        return 0

# Est-ce que le relay c'est allumé puis est maintenant éteind aujourd'hui ? (dans les 12 heures)
def relayUpDownToday(relay):
    req=db.execute('SELECT count(date) FROM relay WHERE relay_number = '+str(relay)+' AND info = "E" AND (valeur = 1 OR valeur = 2) AND date > ' + str(t-720) + ' ORDER BY date DESC LIMIT 2')
    try:
        if req.fetchone()[0] == 2:
            return True
        else:
            return False
    except:
        return False

# Est-ce que le relay c'est allumé aujourd'hui ? (dans les 12 heures)
def relayUpToday(relay):
    #print('SELECT count(date) FROM relay WHERE relay_number = '+str(relay)+' AND info = "E" AND (valeur = 2 OR valeur = 3) AND date > ' + str(t-43200) + ' ORDER BY date DESC LIMIT 1')
    db.execute('SELECT count(date) FROM relay WHERE relay_number = '+str(relay)+' AND info = "E" AND (valeur = 2 OR valeur = 3) AND date > ' + str(t-43200) + ' ORDER BY date DESC LIMIT 1')
    resulta=db.fetchone()[0]
    try:
        if resulta >= 1:
            return True
        else:
            return False
    except:
        return False


def timeUpMax(timeUp):
    # Si le temps 
    if relayLastUpAuto(relayId)+timeUp < t:
        return True
    else:
        return False
		
def timeUpMin(timeUp):
    # Si le temps 
    if relayLastUpAuto(relayId)+timeUp > t:
        return True
    else:
        return False

# ~ @todo : Mode debug in write
# ~ #########################
# ~ # Paramètre lancement script
# ~ #########################
# ~ # HELP @todo à faire !!!
# ~ patternHelp = re.compile("help$")
# ~ if patternHelp.match(sys.argv[1]):
    # ~ print("")
    # ~ print("Usage: python3 "+sys.argv[0]+" debugScript RelayNumber ObjectDataFile")
    # ~ print(" - RelayNumber : 0-9")
    # ~ print(" - jsonDataFile : /tmp/Pvmonit_domo_debug.data (option, sinon c'est téléchargé en direct)")
    # ~ print("")
    # ~ print("Le fichier jsonDataFile contient par exemple : ")
    # ~ print("{'CS': 'Bulk (en charge)', 'P': '-58', 'PPVT': '6', 'SOC': '95.6'}")
    # ~ sys.exit(1)

# ~ # DEBUG
# ~ patternDebug = re.compile("debugScript")
# ~ if patternDebug.match(sys.argv[1]):
    # ~ if len (sys.argv) != 3 :
        # ~ print("Erreur, il manque des arguments pour le debug : ")
        # ~ print("Usage: python3 "+sys.argv[0]+" debugScript RelayNumber ObjectDataFile")
        # ~ print(" - RelayNumber : 0-9 (numéro du script contenu dans relay.script.d)")
        # ~ print(" - jsonDataFile : /tmp/Pvmonit_domo_debug.data (option, sinon c'est téléchargé en direct)")
        # ~ print("")
        # ~ print("Le fichier jsonDataFile contient par exemple : ")
        # ~ print("{'CS': 'Bulk (en charge)', 'P': '-58', 'PPVT': '6', 'SOC': '95.6'}")
        # ~ sys.exit (1)


logMsg(3, "Début de la boucle")
while 1:
    # XML data recup
    t=int(time.time())
    # S'il y a trop d'erreur : 
    if xmlfileCheckError >= configGet('domo', 'fileCheckError'):
        logMsg(1, 'Trop d\'erreur, on patiente 10 secondes')
        time.sleep(10)
        xmlfileCheckError=0
    if xmlLastCheck+configGet('domo', 'dataCheckTime') < t:
        download_data()
        if not os.path.isfile(configGet('tmpFileDataXml')):
            logMsg(2, "Le fichier XML de donnée " + configGet('tmpFileDataXml') + " n'existe pas.")
            xmlfileCheckError=xmlfileCheckError+1
        elif os.path.getmtime(configGet('tmpFileDataXml'))+configGet('domo', 'fileExpir') < t :
            logMsg(2, "Le fichier data est périmé !")
            xmlfileCheckError=xmlfileCheckError+1
        else:
            logMsg(3, "Récupération des données XML (état de l'installation solaire)")
            try:
                # Tentative de lecture
                tree = etree.parse(configGet('tmpFileDataXml'))
                datacount=0
                for datas in tree.xpath("/devices/device/datas/data"):
                    if datas.get("id") in configGet('domo', 'valueUse'):
                        datacount = datacount + 1
                        for data in datas.getchildren():
                                if data.tag == "value":
                                    xmlData[datas.get("id")]=data.text
                logMsg(5, pprint.pprint(xmlData))
                # Test intégrité des données
                controle=configGet('domo', 'valueUse')
                xmlDataControle=True
                for xmlDataVerif in xmlData:
                    pattern = re.compile(controle[xmlDataVerif])
                    if not pattern.match(xmlData[xmlDataVerif]):
                        logMsg(2, "Contrôle données erreur : " + xmlDataVerif + " = " + xmlData[xmlDataVerif])
                        xmlDataControle=False
            except:
                # Si ce n'est pas bon, c'est que les données ne sont pas bonnes ou incomplètes, on télécharge donc de nouveau le fichier XML
                logMsg(1, "Erreur dans la lecture du XML la syntax n'est pas bonne ?")
                xmlDataControle=False

            if xmlDataControle == True:
                xmlLastCheck=t
                xmlDataLastControleOk=t
                xmlfileCheckError=0                    
                heartGo=True
            else:
                # Directement on ce met en mode "trop d'erreur" on patiente
                xmlfileCheckError=configGet('domo', 'fileCheckError')
                
    if xmlDataLastControleOk+configGet('domo', 'xmlDataExpir') < t:
        logMsg(1, "Les données XML ont expirés, on stop le hearbeat")
        heartGo=False

    #########################
    # Le heartbeat
    #########################
    if heartGo == True and heartLastCheck+configGet('domo', 'heartbeatTime') < t:
        writeNumber(int(ord("H")))
        logMsg(5, 'Heardbeat envoyé')
        heartLastCheck=t
        
    # Si les données sont bonnes : 
    if (xmlDataControle == True):
        #########################
        # Data Relay
        #########################
        if relayDataLastCheck+configGet('domo', 'relay', 'dataFreq') < t:
            logMsg(4, 'On récupère les données des relay (via i2c arduino)')
            # A FAIRE
            # Simulation monsieur l'arbitre
            #// Etat : 
            #//  - 0 : off force
            #//  - 1 : off auto
            #//  - 2 : on auto
            #//  - 3 : on force
            #// Mode
            #//  - 0 : Null
            #//  - 1 : Off 
            #//  - 2 : Auto
            #//  - 3 : On
            
            time.sleep(0.3)
            # Requête i2c pour demande de data (état et mode des relay)
            i2cResults = bus.read_i2c_block_data(configGet('domo', 'i2c', 'adress'), int(ord('D')), configGet('domo', 'relay', 'nb')*2+1)
            # On conserve pour comparaison
            try:
                relayEtatOld=relayEtat
                firstDataRece=False
            except:
                firstDataRece=True
            # Remise à 0
            relayEtat=[]
            relayMod=[]
            x=0
            dataOrdre=1
            logEtat=""
            logMod=""
            for i2cDatas in i2cResults:
                # Si les données sont présentes
                if i2cDatas !=  255:
                    if i2cDatas == 29:  # C'est le sépartateur : https://fr.wikibooks.org/wiki/Les_ASCII_de_0_%C3%A0_127/La_table_ASCII
                        dataOrdre=2
                        x=0
                    elif dataOrdre == 1:
                        relayEtat.insert(x,i2cDatas)
                        logEtat=logEtat+","+str(i2cDatas)
                        # Détection des changement pour les mettres en BD
                        if firstDataRece == False:
                            if i2cDatas != relayEtatOld[x]:
                                logMsg(2, 'Un changement est détecté sur le relay ' + str(x) + ', il est passé à '+str(i2cDatas)+' on l\'enregistre en BD !')
                                testDoublon=db.execute('SELECT count(id) FROM relay WHERE relay_number = ' + str(x) + ' AND info = "E" AND valeur = '+str(i2cDatas)+' AND date > ' + str(t-configGet('domo', 'relay', 'dataFreq')) )
                                if testDoublon.fetchone()[0]  == 0:
                                    logInDb(x, 'E', i2cDatas, '')
                                else:
                                    logMsg(2, 'Finalement non, c\'est un doublon')
                        # Premier lancement, on enregistre les init en base
                        if firstDataRece == True:
                            testDoublon=db.execute('SELECT count(id) FROM relay WHERE relay_number = ' + str(x) + ' AND info = "E" AND valeur = '+str(i2cDatas)+' ORDER BY date ASC') 
                            if testDoublon.fetchone()[0]  == 0:
                                logInDb(x, 'E', i2cDatas, 'Init')
                            else:
                                logMsg(2, 'Finalement non, on log pas d\'INIT c\'est un doublon')
                    else:
                        relayMod.insert(x,i2cDatas)
                        logMod=logMod+","+str(i2cDatas)
                        
                    x=x+1
            logMsg(3, "DATA reçu : Etat " + logEtat)
            logMsg(3, "DATA reçu : Mod " + logMod)
            relayDataLastCheck=t
        #########################
        # On joue les scripts
        #########################
        if scriptExecLast+configGet('domo', 'relay', 'scriptExecInterval') < t:
            logMsg(4, 'On joue les script des relay en mode auto')
            relayId=0
            for mod in relayMod:
                scriptFile=configGet('domo','relay', 'scriptDir') + "/" + str(relayId) + ".py"
                if (relayMod[relayId] != 2):
                    logMsg(4, 'Le relay ' + str(relayId) + ' n\'est pas en mode automatique, mais en mode ' + str(relayMod[relayId]))
                elif (relayEtat[relayId] == 0):
                    logMsg(4, 'Les relay sont encore en état 0, les mods sont donc juste changé et les relay n\'on  pas eu le temps de changé d\'état, on patiente')
                elif not os.path.isfile(scriptFile): 
                    logMsg(4, 'Pas de script ' + scriptFile)
                else:
                    # On joue les script 1 par 1, on attend pour jouer le suivant
                    if (relayId > dernierScriptJoue):
                        dernierScriptJoue=relayId
                        logMsg(3, 'Lecture du script ' + scriptFile)
                        returnEtat=None
                        returnLog=None
                        execfile(scriptFile)
                        if returnLog != None:
                            logMsg(1, '['+str(relayId)+'] ' + returnLog)
                        if returnEtat != None and returnEtat != relayEtat[relayId]:
                            logMsg(1, 'Un changement d\'état vers ' + str(returnEtat) + ' de est demandé pour le relay ' + str(relayId))
                            data=[relayId,returnEtat]
                            bus.write_i2c_block_data(configGet('domo', 'i2c', 'adress'), int(ord('O')), data)
                            time.sleep(0.2)
                            sortie=True
                            logInDb(relayId, 'E', returnEtat, returnLog)
                        else:
                            logMsg(4, 'Pas de changement d\'état demandé pour le relay ' + str(relayId))
                # Remise à 0 de la position de lecture des scripts
                if (relayId >= len(relayMod)-1):
                    dernierScriptJoue=-1
                relayId=relayId+1
                # Sortie de la boucle demandé
                if (sortie == True):
                    sortie=False
                    break
            scriptExecLast=t
            
        
    # Pour être gentil avec le système
    time.sleep(0.05)

con.close()

